MM-27936: Allow resizing the channel sidebar and right-hand sidebar

kyeongsoosoo opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->
Previous WebApp PR: https://github.com/mattermost/mattermost-webapp/pull/11864

- make `lhs` and `rhs` resizable
- add `size` property to sidebar store
#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->

Fixes https://github.com/mattermost/mattermost-server/issues/21675

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.

For an easier comparison of UI changes a table (template below) can be used.

|  before  |  after  |
|----|----|
| <insert before screenshot here> | <insert after screenshot here> |

-->

For window widths <900:

not resizable

For window widths 900-1199:


https://user-images.githubusercontent.com/59413880/209613520-4defe4d2-8cd4-47ac-9b3c-56929e8ed2a1.mov


For window widths 1200-1679:

https://user-images.githubusercontent.com/59413880/209613621-411d0279-c16b-4bdf-b430-4130b41d6d43.mov


For window widths >= 1680px:


https://user-images.githubusercontent.com/59413880/209613761-d9d2c49b-144f-4def-b784-1f4c8776ba0b.mov



#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
Allow resizing the channel sidebar and right-hand sidebar
```

mattermost-build commented:
Hello @kyeongsoosoo,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

kyeongsoosoo commented:
@nevyangelova I re-opened the PR and checked that it's working well :)

mattermost-build commented:
E2E tests not automatically triggered, because PR has no approval yet. Please ask a developer to review and then try again to attach the QA label.

matthewbirtch commented:
Is there an issue with spinning up test servers?

AshishDhama commented:
@kyeongsoosoo could you please fix the failing test.

M-ZubairAhmed commented:
/update-branch

kyeongsoosoo commented on webapp/channels/src/stores/local_storage_store.ts:
```diff
@@ -15,6 +15,8 @@ export const getPenultimateChannelNameKey = (userId: string, teamId: string) =>
 const getWasLoggedInKey = () => 'was_logged_in';
 const teamIdJoinedOnLoadKey = 'teamIdJoinedOnLoad';
+const getLhsWidthKey = (userId: string) => ['lhs_width', userId].join(':');
+const getRhsWidthKey = (userId: string) => ['rhs_width', userId].join(':');
```
There was a review in the previous PR, but it seems to be missing.
@esethna Could you check it out?
It is about saving the resize state of the sidebar

![스크린샷 2023-05-01 오후 11 37 52](https://user-images.githubusercontent.com/59413880/235468898-eff0f68f-348c-4dd6-9a77-166381e2e2f2.png)

matthewbirtch replied:
I don't think we need to sync across all clients. As long as the resize state can be maintained on refresh from the same client, I think that would suffice.

kyeongsoosoo replied:
I changed it not to use user_id.

kyeongsoosoo commented:
> @kyeongsoosoo could you please fix the failing test.

@AshishDhama I think it was fixed by updating the branch

kyeongsoosoo commented:
> @kyeongsoosoo could you please fix the failing test.

@AshishDhama 
Circleci test appears to be failing, but I don't have access to view it.
Could you tell me how to see it?

AshishDhama commented on webapp/channels/src/utils/constants.tsx:
```diff
@@ -2029,6 +2036,65 @@ export const WindowSizes = {
     DESKTOP_VIEW: 'desktopView',
 };
 
+export enum SidebarSize {
```
@kyeongsoosoo Could you please move these to separate file.

kyeongsoosoo replied:
I moved these to the sidebar component folder.

AshishDhama commented:
> > @kyeongsoosoo could you please fix the failing test.
> 
> @AshishDhama Circleci test appears to be failing, but I don't have access to view it. Could you tell me how to see it?

I reran the test it passed.

nevyangelova commented:
@kyeongsoosoo a fixed was merged to the mattermost server master branch, could you update your fork ? Thank you :)

AshishDhama commented:
/update-branch

kyeongsoosoo commented on webapp/channels/src/stores/local_storage_store.ts:
```diff
@@ -115,6 +117,22 @@ class LocalStorageStoreClass {
+        return this.getItem(getRhsWidthKey(userId));
+    }
+
+    setRhsWidth(userId: string, width: number) {
```
I changed it not to use user_id.

matthewbirtch commented:
@AshishDhama any idea why the test server won't spin up?

kyeongsoosoo commented on webapp/channels/src/sass/responsive/_tablet.scss:
```diff
@@ -177,7 +177,7 @@
 
         &.sidebar--right--expanded {
-            width: min(60vw, 75rem);
+            width: min(60vw, 75rem) !important;
```
I removed `important` by modifying component logic that was affecting css

AshishDhama commented:
> @AshishDhama any idea why the test server won't spin up?

Mostly due to failing checks,

@kyeongsoosoo have you synced master branch of your fork with latest master.

kyeongsoosoo commented:
/update-branch

AshishDhama commented:
@kyeongsoosoo @matthewbirtch Test server is up

kyeongsoosoo commented:
@matthewbirtch Fixed a problem that could occur when resizing browser width

AshishDhama commented on webapp/channels/src/components/resizable_sidebar/resizable/index.tsx:
```diff
@@ -0,0 +1,317 @@
+    }, [onResizeEnd]);
+
+    useEffect(() => {
+        if (!isResizeLineSelected || !dir || disabled) {
```
@kyeongsoosoo What is the benefit of this if statement?
cc// @M-ZubairAhmed @hmhealey

kyeongsoosoo replied:
To prevent unnecessary event registration, the event was registered at the moment of pressing the resize line

AshishDhama commented:
@kyeongsoosoo Looking good just one question

furqanmlk commented:
/update-branch

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

hmhealey commented:
I removed the stale label since this is waiting on reviews from our side

yasserfaraazkhan commented:
/e2e-test

M-ZubairAhmed commented:
@calebroseland gentle ping on this one 😄

yasserfaraazkhan commented:
Hi @M-ZubairAhmed 
this change is not working on the spinwick, can you help in verifying ?

calebroseland reviewed (changes requested):
Hi @kyeongsoosoo, I've attempted to review this a few times now, and I keep running into mental blocks of my own making. I think what I've been running against is a difference in approach and paradigm.

Your PR addresses quite a number of UX details, and your attention to detail there is fantastic. At the same time, there's quite a lot of code and infrastructure, and it's a very JS-oriented approach.

`Resizable` is pretty well-written and handles quite a few details, but there're still a lot of layout effects and other event hooks and default props—some absolutely necessary, for sure, but there are a good number of dependencies and conditionals—onLimitChange, min/max, initialWidth vs. defaultWidth, etc. There's also unnecessary memoization in a number of places over in `ResizableLhs`/`ResizableRhs`

When we overhauled the main layout last year (https://github.com/mattermost/mattermost-webapp/pull/11019), we used a very CSS-based paradigm and approach to the main layout and responsiveness.

Back to resizing: traditionally or conventionally, directly overriding an element's `style.width` property is often used. For simple cases, this is absolutely fine, but one big caveat is that is can bypass any number of CSS rules in the process—that can be a problem in intricate scenarios, sometimes requiring emulating responsive and other CSS-based logic in JS (possible, but hard to do).

#### Proposal
I wanted to submit the idea of pivoting a bit to use simpler infrastructure, with a CSS-based approach using CSS Vars to provide an override value to the same CSS rules that normally define the respective elements `width`s. 

This does a few things IMHO:

1. Layout, styling, and min/max logic, and responsiveness all (generally) stay within CSS, where they fit most naturally.
2. Event listeners, preference storage, and other misc things generally stay within JS.

As an example, https://github.com/mattermost/mattermost/pull/23736 may explain better. It's 
based on parts of your PR. It's certainly not complete, and missing some details that are included in your PR, but it does condense a decent amount of the infrastructure into their respective areas (fewer files changed). 

Please feel free to take this and do as you see fit with it. Let me know what you think, and thanks again for all your work on this!

calebroseland commented:
Also found some quirks/edge cases while testing:

https://github.com/mattermost/mattermost/assets/11724372/8d196378-c2b0-4254-a811-00eb1e01e1b5

yasserfaraazkhan commented:
/e2e-test

mattermost-build commented:
The command /e2e-test requires all PR checks to pass.

yasserfaraazkhan commented:
/update-branch

mattermost-build commented:
Error trying to update the PR.
Please do it manually.

kyeongsoosoo commented:
@calebroseland Thanks for the review. I agree that the js-oriented approach has a caveat.
I think it'd be good to approach it with a css-based way. I'll try to apply the infrastructure.

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

johndavidlugtu commented:
@kyeongsoosoo friendly ping here -- let us know if you have some questions for the team

kyeongsoosoo commented:
@calebroseland 
I changed the code to use a css-based approach.
I think it was a good choice as the code got cleaner.
I also fixed other reviews, including fixing edge cases.
Sorry for the delay🥲

matthewbirtch commented:
I'll review this again once we get a test server up and running. Seems like it's not generating ATM

larkox commented:
/update-branch

matthewbirtch reviewed (changes requested):
Took another look at this and it's working great. A couple of minor things I noticed:

1. When double-clicking to reset the RHS width to the default, can we remove the animation and just make it immediately go to the default width without any transition?
2. For browser window widths between 769-900px, I believe we spec'd it to not be resizable at all, but we should allow resizing to the minimum of 200px and max 264px still I think at these widths. I can update the spec to reflect this
3. There is something a little odd happening when opening the RHS. While on a large screen, I resized the RHS to it's max-width. Then I closed the RHS and reduced the window size to a smaller width. Up re-opening the RHS it first opens wide and after a short delay collapses its width to the proper smaller width (see video below).

https://github.com/mattermost/mattermost/assets/2040554/0dc02a61-5192-42bf-a7eb-fd5ec9c7c613

kyeongsoosoo commented:
/update-branch

kyeongsoosoo commented:
> Took another look at this and it's working great. A couple of minor things I noticed:

@matthewbirtch I've updated based on the reviews☺️

matthewbirtch reviewed (approved):
Awesome stuff @kyeongsoosoo!

mattermost-build commented:
E2E tests not automatically triggered, because the PR is not in a mergeable state. Please update the branch with the base branch and resolve outstanding conflicts.

harshilsharma63 commented:
Damn! I love this PR 😍😍😍😍 Thanks for making this happen!

harshilsharma63 commented:
/update-branch

harshilsharma63 commented:
I've fixed the part where the widths were scoped to a single team. Now every team shares the same width.

kyeongsoosoo commented:
> I've fixed the part where the widths were scoped to a single team. Now every team shares the same width.

Thanks for the review😄

yasserfaraazkhan commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/364883) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/364883)

calebroseland commented on webapp/channels/src/components/resizable_sidebar/resizable_divider.tsx:
```diff
@@ -100,7 +100,7 @@ function ResizableDivider({
     const currentUserID = useSelector(getCurrentUserId);
 
     const [isActive, setIsActive] = useState(false);
-    const [width, setWidth] = useGlobalState<number | null>(null, `resizable_${name}:${currentUserSuffix}`, currentUserID);
```
this was a selector function, not a string, so incompatible as-used; removed as suffixed by 3rd param/currentUserID

calebroseland reviewed (approved):
I still question the need for the extra event listeners and new/separate breakpoints, but this is getting to be good enough and it could possibly be iterated on in the future.

I was seeing some responsive issues with the LHS header—at it's narrowest it would cut off the `+` menu, so I pushed a change to address that:

| before | after |
|---|---|
|  ![CleanShot 2023-08-24 at 12 07 14](https://github.com/mattermost/mattermost/assets/11724372/10a37de0-646e-4d9c-ac9e-814e1ef24e32) |  ![CleanShot 2023-08-28 at 17 37 23](https://github.com/mattermost/mattermost/assets/11724372/8394d78b-3e60-4aee-b4d0-83d158a34b31) |


There's still one odd UX bit—not critical IMO, but at viewport widths less than `900` wide (but still over the mobile threshold of `768`), the resize handle doesn't work (but still grab-able) and it snaps back to its default size if it was previously resized.


https://github.com/mattermost/mattermost/assets/11724372/e921e54c-35e2-4276-bc62-648f3f012d43

 Nonetheless, this is a fantastic piece of work and will be really nice to get this in! 
 
 Thanks, @kyeongsoosoo!

M-ZubairAhmed commented on webapp/channels/src/components/advanced_text_editor/formatting_bar/hooks.tsx:
```diff
@@ -18,23 +18,24 @@ export function useGetLatest<T>(val: T) {
+        if (ref.current.clientWidth < 310) {
+            setWideMode('min');
+        }
+    }, 10), [ref.current?.clientWidth]);
```
[ref's should not](https://epicreact.dev/why-you-shouldnt-put-refs-in-a-dependency-array/) be used as dependency

kyeongsoosoo replied:
Thanks for the link. I fixed it.

M-ZubairAhmed commented on webapp/channels/src/components/resizable_sidebar/resizable_divider.tsx:
```diff
@@ -0,0 +1,267 @@
+        reset();
+        setWidth(null);
+    };
+
```
nit : return null when disabled

kyeongsoosoo replied:
Done!

M-ZubairAhmed commented on webapp/channels/src/components/window_size_observer/WindowSizeObserver.tsx:
```diff
@@ -0,0 +1,93 @@
+        };
+    }, [handleSidebarMediaQueryChangeEvent, setSidebarSizeWhenWindowResized, updateSidebarSize]);
+
+    return <></>;
```
nit can we return null instead of empty jsx?

kyeongsoosoo replied:
I changed it to return null.

M-ZubairAhmed reviewed (commented):
Looks and works great. Just some nits

M-ZubairAhmed reviewed (approved):
This looks good :shipit: Thank you

marianunez commented:
@kyeongsoosoo thank you for this great contribution and your patience working through the feedback! This is a high-impact feature 🚀
